### ISSUE TITLE
Revisit FileBasedSignatureProvider for a possible performance optimization

### ISSUE THREAD

**Issue by imback82** [microsoft/hyperspace#271]
could you also hide this feature behind a flag which defaults to false? 
My concern is I think this would be unnecessary calculation for regular scenarios. In any case, if signature doesn't match, RuleUtils will make sure that if index can be used, it will be used right? I am just wondering why bother O(NlogN) for sorting always for a rare scenario which will still work without this change(assuming hybrid scan is enabled eventually by default)?

Alternately if we still want to fix this issue, could you avoid sorting? Instead you could convert fileInfos to set and then create fingerprint. Set will ensure the order of iteration is unique for a unique collection of elements. That way we can still achieve O(N).

cc @imback82

_Originally posted by @apoorvedave1 in https://github.com/microsoft/hyperspace/issues/268#issuecomment-732451580_

**Comment by imback82** [comment]
Related thread: https://github.com/microsoft/hyperspace/pull/268#discussion_r529071399

**Comment by sezruby** [comment]
I think we cannot guarantee the order of HashSet iterator. It can differ depending on the internal implementation. (e.g. different hash bucket size, hash functions..)

**Comment by clee704** [comment]
Maybe I misunderstood the issue, but why is the order important? Why can't we simply XOR or even just add signatures?

Also, is there a specific reason we use String instead of byte[] and use md5Hex instead of md5 for signatures?

**Comment by sezruby** [comment]
@clee704 Yea I'm good with XOR approach & using byte[] (be aware of its format in JSON)

cc @imback82

**Comment by clee704** [comment]
Yeah, we should do proper HEX encoding/decoding when JSON is involved.

By the way, how about changing the signature from MD5 to SHA1 for a little more bits (128 vs. 160), and ultimately, shouldn't we make it easy to switch the underlying hashing algorithm?

**Comment by sezruby** [comment]
Btw changing signature will break backward compatibility; so would be good to go with other breaking changes.

**Comment by clee704** [comment]
I assume that the signatures are stored as JSON in users' data lake. Therefore, any change to the signature computation will  make existing signatures won't work. Am I right? Then how could you merge the PR #268? @sezruby 

Do we provide some migration utility to update their index data?

**Comment by rapoth** [comment]
@clee704 Migration utility is a great idea! Can you please open a feature request and link ot here so we can keep track of it? While it may not be necessary for the current version, I think as we move towards stabilization, it'd be useful to provide one for users so we don't always have to worry about breaking changes.

**Comment by clee704** [comment]
@rapoth [Release notes](https://github.com/microsoft/hyperspace/releases) state that whenever there are breaking changes, indexes should be "reconstructed". Does it mean that users should manually create indexes again? Because refreshAction won't work because it can't load the previous version's IndexLogEntry, or weird things can happen if it can (e.g. only the signature computation logic changed). Right? So that could be the why we might need a migration utility?

**Comment by sezruby** [comment]
>Does it mean that users should manually create indexes again?

Yes
> Because refreshAction won't work because it can't load the previous version's IndexLogEntry, or weird things can happen if it can (e.g. only the signature computation logic changed). Right?

Right, usually it failed to create IndexLogEntry, but in that case we could reuse the previous index data. 
I think we could catch the exception from building IndexLogEntry & trigger the utility to create a new log file.

**Comment by clee704** [comment]
> > Does it mean that users should manually create indexes again?
> 
> Yes
> 

Hmm, okay. Now I see a need for a migration utility.

> > Because refreshAction won't work because it can't load the previous version's IndexLogEntry, or weird things can happen if it can (e.g. only the signature computation logic changed). Right?
> 
> Right, usually it failed to create IndexLogEntry, but in that case we could reuse the previous index data.
> I think we could catch the exception from building IndexLogEntry & trigger the utility to create a new log file.

Blindly trying to create a log from JSON and catching an exception doesn't seem to be a good approach, because we can't be sure that the constructed log is valid. For example, if only the signature computation logic is updated between Hyperspace versions, the log will be created just fine but with invalid signatures.
